stage-1: preparation

MariaKonstantinov opened the pull request:
Stage-1: Preparation: set up initial files

adzhenalieva commented:
# Summary:
## Backend

### Performance Criteria
- [ ] <!--3.64--> The repository contains all the necessary infrastructure files:
  - [x] `package.json`;
  - [x] `.editorconfig`;
  - [x] `.eslintrc` is required for extending the configuration of airbnb-base
		- [x] `package.json` contains the `devDependencies` needed for the linter to work correctly
		- [ ] An exception for `_id` is added
		   _https://snipboard.io/J9vtMy.jpg there are linter errors that need to be fixed_
		- [x] The following rules are forbidden: `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line`
  - [x] `.gitignore` should contain at least the `node_modules` folder
- [ ] <!--3.64--> There are no lint errors. When running `npx eslint .` they should be absent.
   _https://snipboard.io/GNrPQt.jpg The linter settings must contain only an exception for an `_id` and next. The other rules shouldn't be overridden_
- [x] <!--3.64--> The scripts section of the package.json file contains the following:
  - [x] An `npm run start` command that starts the server on `localhost:3000`.
  - [x] An `npm run dev` command that starts the server on `localhost:3000` with hot reloading.
- [ ] <!--21.81--> The following routes function as described:
  - [x] A request to `GET /users/me` returns information about the user (email and name).
  - [x] `POST /signup` creates a user with the data passed inside the request body.
  - [ ] `POST /signin` returns a JWT when the correct email and password are passed in the request body.
    _In the login controller the JWT_SECRET is undefined_
  - [x] `GET /articles` returns all articles saved by the user.
  - [x] `POST /articles` creates an article with the data passed inside the request body.
  - [x] `DELETE /articles/:articleId` deletes the saved article using `_id`.
- [x] <!--3.64--> Users can't delete saved articles from other user profiles.
- [x] <!--3.64--> All routes are protected with authorization, except for `/signin` and `/signup`.
- [x] <!--3.64--> User routes and article routes are described in separate files.
- [x] <!--7.28--> API errors are handled:
  - [x] Error status codes are used: 400, 401, 403, 404, 409, 500.
  - [x] If something is wrong with the request, the server returns a response with an error message and a corresponding status.
  - [x] The error message matches its type.
  - [x] Asynchronous handlers end with a `catch()` block.
  - [x] The API does not return standard database or Node.js errors.
- [x] <!--3.64--> Safe password storage has been implemented:
  - [x] Passwords are stored in a hashed form.
  - [x] The API does not return a password hash to the client.
- [x] <!--7.28--> Validation implemented correctly:
  - [x] Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
  - [x] Data is validated before being added to the database.
- [x] <!--7.28--> In production mode, the database address is taken from `process.env`.
- [ ] <!--3.64--> The server can be accessed via HTTPS using the domain specified in README.md.
   _There is no link to the server in the README file_
- [ ] <!--7.28--> Storing the private key for creating a JWT is implemented correctly:
  - [ ] For the production build, it is stored in a `.env` file, and this file should not be added to Git.
    _The point will be given after the server check_
  - [ ] In development mode (`process.env.NODE_ENV !== 'production'`), the code runs and works fine, and an error won't occur if there is no `.env` file present.
    _In the login controller the JWT_SECRET is undefined_

### Best Practices
- [x] <!--2.14--> All routes are connected to the `index.js` file, which is located in the `routes` folder, and `app.js` contains one main route handled by routes.
- [x] <!--2.14--> Asynchronous operations are implemented using promises or async/await.
- [x] <!--2.14--> Validation is described in a separate module.
- [x] <!--2.14--> Logging is set up:
  - [x] All requests and responses are logged to the `request.log` file.
  - [x] All errors are logged to the `error.log` file.
  - [x] Log files aren't added to the git repository.
- [x] <!--2.14--> Errors are handled by a centralized handler.
- [x] <!--2.14--> Centralized error handling is described inside a separate module.
- [ ] <!--2.14--> The application API is located in an /api subdirectory or on an api. subdomain (not just name.zone)
  - [ ] Correct:  domain-name.tk/api or api.domain-name.tk
  - [ ] Incorrect: domain-name.tk


### Recommendations
- [x] <!--1.0--> For API errors, classes have been created to extend the Error constructor.
- [x] <!--1.0--> The Helmet module is used to set security-related headers.
- [x] <!--1.0--> Configuration and constants are stored in separate files:
  - [x] The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
  - [x] Application constants (response and error messages) are stored inside a separate file with constants.
- [x] <!--1.0--> A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
- [x] <!--1.0--> The rate limiter is configured in a separate file and imported into `app.js`.

## Number of points: 58

## Additional comments:
-

MariaKonstantinov commented:
Thank you very much for your review.

I wanted to clarify for myself one thing here before I resubmit my project: "The application API is located in an /api subdirectory or on an api. subdomain (not just name.zone); Correct: domain-name.tk/api or api.domain-name.tk, Incorrect: domain-name.tk" - what needs to be fixed here? I looked up into the requirements, and seems they didn't change since the previous project regarding the naming conventions... In my VM's NGINX configuration file I have set: news-explorer.click, api.news-explorer.click, and www.news-explorer.click. 

Thank you in advance!

adzhenalieva commented:
> Thank you very much for your review.
> 
> I wanted to clarify for myself one thing here before I resubmit my project: "The application API is located in an /api subdirectory or on an api. subdomain (not just name.zone); Correct: domain-name.tk/api or api.domain-name.tk, Incorrect: domain-name.tk" - what needs to be fixed here? I looked up into the requirements, and seems they didn't change since the previous project regarding this... In my VM's NGINX configuration file I have set: news-explorer.click, api.news-explorer.click, and [www.news-explorer.click](http://www.news-explorer.click).
> 
> Thank you in advance!

Hello, Maria! I did not see your comment, I am sorry for the delay. The backend part of the project should have 'api' at the beginning of the domain. According to your description, your settings are correct

MariaKonstantinov commented:
Hello, Aisalkyn! All good, thank you very much for your reply. Will go ahead with all the corrections.

MariaKonstantinov commented:
Dear Aisalkyn,

I submitted my code in the same pull request here, after 1st review. Regarding the .eslintrc rules: I can't remove an exception for "_doc" because I am using it in my project the same way I use "_id", particularly in usersController.js. These rules were applied the same way in my previous project (Project 15). I tried to fix everything according to your review. Thank you very much in advance!

adzhenalieva commented:
# Summary:
## Backend

### Performance Criteria
- [ ] <!--3.64--> The repository contains all the necessary infrastructure files:
  - [x] `package.json`;
  - [x] `.editorconfig`;
  - [ ] `.eslintrc` is required for extending the configuration of airbnb-base
		- [x] `package.json` contains the `devDependencies` needed for the linter to work correctly
		- [ ] An exception for `_id` is added
		     _You may use the exception for the `_doc` too, but other exceptions were not deleted: https://snipboard.io/pDMFnz.jpg_
		- [x] The following rules are forbidden: `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line`
  - [x] `.gitignore` should contain at least the `node_modules` folder
- [x] <!--3.64--> There are no lint errors. When running `npx eslint .` they should be absent.
- [x] <!--3.64--> The scripts section of the package.json file contains the following:
  - [x] An `npm run start` command that starts the server on `localhost:3000`.
  - [x] An `npm run dev` command that starts the server on `localhost:3000` with hot reloading.
- [x] <!--21.81--> The following routes function as described:
  - [x] A request to `GET /users/me` returns information about the user (email and name).
  - [x] `POST /signup` creates a user with the data passed inside the request body.
  - [x] `POST /signin` returns a JWT when the correct email and password are passed in the request body.
  - [x] `GET /articles` returns all articles saved by the user.
  - [x] `POST /articles` creates an article with the data passed inside the request body.
  - [x] `DELETE /articles/:articleId` deletes the saved article using `_id`.
- [x] <!--3.64--> Users can't delete saved articles from other user profiles.
- [x] <!--3.64--> All routes are protected with authorization, except for `/signin` and `/signup`.
- [x] <!--3.64--> User routes and article routes are described in separate files.
- [x] <!--7.28--> API errors are handled:
  - [x] Error status codes are used: 400, 401, 403, 404, 409, 500.
  - [x] If something is wrong with the request, the server returns a response with an error message and a corresponding status.
  - [x] The error message matches its type.
  - [x] Asynchronous handlers end with a `catch()` block.
  - [x] The API does not return standard database or Node.js errors.
- [x] <!--3.64--> Safe password storage has been implemented:
  - [x] Passwords are stored in a hashed form.
  - [x] The API does not return a password hash to the client.
- [x] <!--7.28--> Validation implemented correctly:
  - [x] Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
  - [x] Data is validated before being added to the database.
- [x] <!--7.28--> In production mode, the database address is taken from `process.env`.
- [x] <!--3.64--> The server can be accessed via HTTPS using the domain specified in README.md.
- [x] <!--7.28--> Storing the private key for creating a JWT is implemented correctly:
  - [x] For the production build, it is stored in a `.env` file, and this file should not be added to Git.
  - [x] In development mode (`process.env.NODE_ENV !== 'production'`), the code runs and works fine, and an error won't occur if there is no `.env` file present.

### Best Practices
- [x] <!--2.14--> All routes are connected to the `index.js` file, which is located in the `routes` folder, and `app.js` contains one main route handled by routes.
- [x] <!--2.14--> Asynchronous operations are implemented using promises or async/await.
- [x] <!--2.14--> Validation is described in a separate module.
- [x] <!--2.14--> Logging is set up:
  - [x] All requests and responses are logged to the `request.log` file.
  - [x] All errors are logged to the `error.log` file.
  - [x] Log files aren't added to the git repository.
- [x] <!--2.14--> Errors are handled by a centralized handler.
- [x] <!--2.14--> Centralized error handling is described inside a separate module.
- [x] <!--2.14--> The application API is located in an /api subdirectory or on an api. subdomain (not just name.zone)
  - [x] Correct:  domain-name.tk/api or api.domain-name.tk
  - [x] Incorrect: domain-name.tk


### Recommendations
- [x] <!--1.0--> For API errors, classes have been created to extend the Error constructor.
- [x] <!--1.0--> The Helmet module is used to set security-related headers.
- [x] <!--1.0--> Configuration and constants are stored in separate files:
  - [x] The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
  - [x] Application constants (response and error messages) are stored inside a separate file with constants.
- [x] <!--1.0--> A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
- [x] <!--1.0--> The rate limiter is configured in a separate file and imported into `app.js`.

## Number of points: 96

## Additional comments:
- Please, do not forget to update the code on the deployed server, it returns the error from the first review
